File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        
        def recur(d):
            while d>0 and year[d]==False: d-=1
            if d<=0: return 0
            if dp[d]!=-1: return dp[d]
            dp[d] = min(costs[0]+recur(d-1), costs[1]+recur(d-7), costs[2]+recur(d-30))
            return dp[d]
        
        year = [False]*366
        for i in days: year[i]=True
        dp = [-1]*366
        return recur(365)

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.java
class Solution {
    int[][] memo;
    public int mincostTickets(int[] days, int[] costs) {
        memo = new int[days.length][366];
        for(int[] m1: memo) Arrays.fill(m1, -1);
        
        return help(0, 0, days, costs);
    }
    private int help(int i, int freeTill, int[] days, int[] costs){
        if(i==days.length) return 0;
        
        if(memo[i][freeTill] != -1) return memo[i][freeTill];
        
        int ans = Integer.MAX_VALUE;
        
        if(freeTill>=days[i]) ans = Math.min(ans, help(i+1, freeTill, days, costs));
        else{
            ans = Math.min(ans, costs[0]+help(i+1, 0,days, costs));
            
            int newFreeTill = Math.min(days[i]+6, 365);
            ans = Math.min(ans, costs[1]+help(i+1, newFreeTill, days, costs));
            
            newFreeTill = Math.min(days[i]+29, 365);
            ans = Math.min(ans, costs[2]+help(i+1, newFreeTill, days, costs));
        }
        memo[i][freeTill] = ans;
        return memo[i][freeTill];
    }
}